File: tests/cpp/test_protocol.cpp
#include <iostream>
#include <vector>
#include <string>
#include <cstring>
#include "edlora.h"
#include "edlora_crypto.h"

using namespace edlora;

int tests_passed = 0;
int tests_failed = 0;

#define ASSERT_TRUE(condition, name) \
    if (condition) { \
        std::cout << "[PASS] " << name << std::endl; \
        tests_passed++; \
    } else { \
        std::cout << "[FAIL] " << name << " (Line " << __LINE__ << ")" << std::endl; \
        tests_failed++; \
    }

void test_basic_packing() {
    Packet tx;
    tx.sender_id = 0x10;
    tx.receiver_id = 0x20;
    tx.msg_type = MsgType::HEARTBEAT;
    tx.seq_num = 5;
    
    uint8_t buffer[256];
    int size = Protocol::pack(tx, buffer, sizeof(buffer));
    
    ASSERT_TRUE(size == 8, "Basic Packing Size (0 payload)");
    ASSERT_TRUE(buffer[0] == SYNC_BYTE, "Sync Byte OK");
    
    Packet rx;
    bool success = Protocol::unpack(buffer, size, rx);
    ASSERT_TRUE(success, "Basic Unpacking Success");
    ASSERT_TRUE(rx.sender_id == 0x10, "Sender ID Match");
}

void test_string_payload() {
    Packet tx;
    tx.set_payload_string("TEST STRING");
    
    uint8_t buffer[256];
    int size = Protocol::pack(tx, buffer, sizeof(buffer));
    
    Packet rx;
    Protocol::unpack(buffer, size, rx);
    
    char out[256];
    rx.get_payload_string(out, sizeof(out));
    ASSERT_TRUE(std::string(out) == "TEST STRING", "String Payload Match");
}

void test_crc_failure() {
    Packet tx;
    tx.set_payload_string("CORRUPT ME");
    
    uint8_t buffer[256];
    int size = Protocol::pack(tx, buffer, sizeof(buffer));
    
    buffer[7] ^= 0xFF; // Flip bits in payload
    
    Packet rx;
    bool success = Protocol::unpack(buffer, size, rx);
    ASSERT_TRUE(!success, "CRC Failure Detected");
}

void test_crypto_wrapper() {
    Packet tx;
    tx.set_payload_string("SECRET MESSAGE");
    
    crypto::XorCipher cipher(0xAA);
    cipher.process(tx);
    
    uint8_t buffer[256];
    int size = Protocol::pack(tx, buffer, sizeof(buffer));
    
    Packet rx;
    Protocol::unpack(buffer, size, rx);
    
    cipher.process(rx); // Decrypt
    
    char out[256];
    rx.get_payload_string(out, sizeof(out));
    ASSERT_TRUE(std::string(out) == "SECRET MESSAGE", "Crypto Wrapper Match");
}

void test_stream_fragmentation() {
    Packet tx;
    tx.sender_id = 0xAA;
    tx.set_payload_string("FULL");
    
    uint8_t valid_bytes[256];
    int valid_size = Protocol::pack(tx, valid_bytes, sizeof(valid_bytes));
    
    // Simulate HALFPACKET + FULLPACKET + HALFPACKET
    std::vector<uint8_t> stream;
    
    // Insert half packet
    for (int i=0; i<4; i++) stream.push_back(valid_bytes[i]);
    
    // Insert full packet
    for (int i=0; i<valid_size; i++) stream.push_back(valid_bytes[i]);
    
    // Insert half packet
    for (int i=0; i<5; i++) stream.push_back(valid_bytes[i]);
    
    // Scanning Logic
    bool recovered = false;
    Packet rx;
    for (size_t i = 0; i < stream.size(); i++) {
        if (stream[i] == SYNC_BYTE) {
            // Attempt unpack
            if (Protocol::unpack(&stream[i], stream.size() - i, rx)) {
                recovered = true;
                break;
            }
        }
    }
    
    ASSERT_TRUE(recovered, "Recovered from fragmented stream");
    
    char out[256];
    rx.get_payload_string(out, sizeof(out));
    ASSERT_TRUE(std::string(out) == "FULL", "Fragmented Payload Intact");
}

int main() {
    std::cout << "=== Running edLoRa C++ Unit Tests ===" << std::endl;
    
    test_basic_packing();
    test_string_payload();
    test_crc_failure();
    test_crypto_wrapper();
    test_stream_fragmentation();
    
    std::cout << "---" << std::endl;
    std::cout << "Passed: " << tests_passed << ", Failed: " << tests_failed << std::endl;
    
    return (tests_failed == 0) ? 0 : 1;
}

File: tests/python/test_protocol.py
import unittest
import struct
from edlora import Packet, MsgType
from edlora.crypto import XorCipher

class TestEdLoraProtocol(unittest.TestCase):
    def test_basic_packing(self):
        p = Packet(sender_id=0x10, receiver_id=0x20, msg_type=MsgType.HEARTBEAT, seq_num=5)
        packed = p.pack()
        self.assertEqual(len(packed), 8) # 6 header + 0 payload + 2 CRC
        self.assertEqual(packed[0], Packet.SYNC_BYTE)

    def test_string_payload(self):
        p = Packet()
        p.set_payload_string("TEST")
        self.assertEqual(p.payload_len, 4)
        
        packed = p.pack()
        rx_p = Packet.unpack(packed)
        self.assertEqual(rx_p.get_payload_string(), "TEST")

    def test_payload_too_large(self):
        p = Packet()
        p.payload = b'A' * 250
        with self.assertRaises(ValueError):
            p.pack()

    def test_crc_failure(self):
        p = Packet(payload=b"VALID")
        packed = bytearray(p.pack())
        packed[6] = 0xFF # Corrupt a byte in the payload
        with self.assertRaises(ValueError) as context:
            Packet.unpack(bytes(packed))
        self.assertTrue("CRC" in str(context.exception))

    def test_invalid_sync_byte(self):
        p = Packet()
        packed = bytearray(p.pack())
        packed[0] = 0x00 # Invalid sync byte
        with self.assertRaises(ValueError):
            Packet.unpack(bytes(packed))

    def test_crypto_wrapper(self):
        original = Packet(payload=b"SECRET")
        cipher = XorCipher(0xAB)
        
        # Encrypt
        encrypted = cipher.process(original)
        self.assertNotEqual(encrypted.payload, b"SECRET")
        
        # Test serialization of encrypted
        packed = encrypted.pack()
        
        # Decrypt
        rx = Packet.unpack(packed)
        decrypted = cipher.process(rx)
        
        self.assertEqual(decrypted.payload, b"SECRET")

    def test_stream_fragmentation(self):
        # Simulate: HALFPACKET + FULLPACKET + HALFPACKET
        p = Packet(sender_id=0xAA, msg_type=MsgType.COMMAND)
        p.set_payload_string("FULL")
        valid_packet_bytes = p.pack()
        
        # Craft a stream with garbage/partial packets on both ends
        half_packet_before = valid_packet_bytes[:4]
        half_packet_after = valid_packet_bytes[:5]
        stream = half_packet_before + valid_packet_bytes + half_packet_after
        
        # Stream parser scanning logic
        recovered = None
        for i in range(len(stream)):
            if stream[i] == Packet.SYNC_BYTE:
                try:
                    # Try unpacking from this sync byte onwards
                    rx = Packet.unpack(stream[i:])
                    recovered = rx
                    break
                except ValueError:
                    pass # Keep scanning
        
        self.assertIsNotNone(recovered)
        self.assertEqual(recovered.get_payload_string(), "FULL")

if __name__ == '__main__':
    unittest.main()
